### PR TITLE
Bug 1955490: Thanos ruler Statefulsets should have 2 replicas and hard affinity set

### PR DESCRIPTION
**Pull request by raptorsun**
* [ ] I added CHANGELOG entry for this change.
* [x] No user facing changes, so no entry in CHANGELOG was needed.

Changes:
1. Add hard pod anti-affinity to thanos-ruler.
2. Thanos-ruler configuration uses ``kube-thanos-rule.libsonnet`` from upstream kube-thanos project

2 replica is not guaranteed due to [lack of MaxUnavailable Rolling Update  in Stateful Set](https://github.com/kubernetes/kubernetes/issues/68397). There is already [a PR on Kubernetes](https://github.com/kubernetes/kubernetes/pull/82162), once it is merged we can come back to this issue.

### CODE REVIEW

File: jsonnet/thanos-ruler.libsonnet [hunk 1]
```diff
@@ -1,457 +1,458 @@
 // TODO(paulfantom): This should use upstream kube-thanos project.
```
Comment (simonpasquier) [inline jsonnet/thanos-ruler.libsonnet:1]: should we remove this line?

Comment (raptorsun) [reply jsonnet/thanos-ruler.libsonnet:1]: Absolutely. This is already done by pull request

File: jsonnet/thanos-ruler.libsonnet [hunk 2]
```diff
@@ -1,457 +1,458 @@
...
-  serviceMonitor: {
-    apiVersion: 'monitoring.coreos.com/v1',
-    kind: 'ServiceMonitor',
-    metadata: {
-      name: 'thanos-ruler',
-      namespace: cfg.namespace,
-      labels: {
-        'app.kubernetes.io/name': 'thanos-ruler',
+      spec+: {
+        selector: cfg.selectorLabels,
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:252]: we can inherit this from the kube-thanos lib: https://github.com/thanos-io/kube-thanos/blob/main/jsonnet/kube-thanos/kube-thanos-rule.libsonnet#L85

File: jsonnet/thanos-ruler.libsonnet [hunk 3]
```diff
@@ -1,457 +1,458 @@
...
+        sessionAffinity: 'ClientIP',
+        type: 'ClusterIP',
+        clusterIP:: 'None',
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:255]: we can inherit this from upstream: https://github.com/thanos-io/kube-thanos/blob/main/jsonnet/kube-thanos/kube-thanos-rule.libsonnet#L84

File: jsonnet/thanos-ruler.libsonnet [hunk 4]
```diff
@@ -1,457 +1,458 @@
...
-        'config.openshift.io/inject-trusted-cabundle': 'true',
+    trustedCaBundle: {
+      apiVersion: 'v1',
+      kind: 'ConfigMap',
+      metadata: {
+        name: 'thanos-ruler-trusted-ca-bundle',
+        namespace: cfg.namespace,
+        labels: {
+          'config.openshift.io/inject-trusted-cabundle': 'true',
+        },
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:32]: we should also add the commonLabels: https://github.com/thanos-io/kube-thanos/blob/main/jsonnet/kube-thanos/kube-thanos-rule.libsonnet#L29-L34. They should be available as `cfg.commonLabels`.

File: jsonnet/thanos-ruler.libsonnet [hunk 5]
```diff
@@ -1,457 +1,458 @@
...
-      name: 'thanos-ruler',
-      namespace: cfg.namespace,
-    },
-    spec: {
-      to: {
-        kind: 'Service',
+    route: {
+      apiVersion: 'v1',
+      kind: 'Route',
+      metadata: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:42]: we can add the commonLabels to the metadata

File: jsonnet/thanos-ruler.libsonnet [hunk 6]
```diff
@@ -1,457 +1,458 @@
...
-      name: 'thanos-ruler-monitoring',
-    },
-    roleRef: {
-      apiGroup: 'rbac.authorization.k8s.io',
-      kind: 'ClusterRole',
-      name: 'cluster-monitoring-view',
+    clusterRoleBindingMonitoring: {
+      apiVersion: 'rbac.authorization.k8s.io/v1',
+      kind: 'ClusterRoleBinding',
+      metadata: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:108]: we can add commonLabels here

File: jsonnet/thanos-ruler.libsonnet [hunk 7]
```diff
@@ -1,457 +1,458 @@
...
-      namespace: cfg.namespace,
-      labels: {
-        'app.kubernetes.io/name': 'thanos-ruler',
+    grpcTlsSecret: {
+      apiVersion: 'v1',
+      kind: 'Secret',
+      metadata: {
+        name: 'thanos-ruler-grpc-tls',
+        namespace: cfg.namespace,
+        labels: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:129]: replace by commonLabels

File: jsonnet/thanos-ruler.libsonnet [hunk 8]
```diff
@@ -1,457 +1,458 @@
...
-        'app.kubernetes.io/name': 'thanos-ruler',
+    // holds the secret which is used encrypt/decrypt cookies
+    // issued by the oauth proxy.
+    oauthCookieSecret: {
+      apiVersion: 'v1',
+      kind: 'Secret',
+      metadata: {
+        name: 'thanos-ruler-oauth-cookie',
+        namespace: cfg.namespace,
+        labels: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:145]: replace by commonLabels

File: jsonnet/thanos-ruler.libsonnet [hunk 9]
```diff
@@ -1,457 +1,458 @@
...
+    // holds the htpasswd configuration
+    // which includes a static secret used to authenticate/authorize
+    // requests originating from grafana.
+    oauthHtpasswdSecret: {
+      apiVersion: 'v1',
+      kind: 'Secret',
+      metadata: {
+        name: 'thanos-ruler-oauth-htpasswd',
+        namespace: cfg.namespace,
+        labels: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:162]: replace by commonLabels

File: jsonnet/thanos-ruler.libsonnet [hunk 10]
```diff
@@ -1,457 +1,458 @@
...
+
+    // query config which holds http configuration
+    // for communication between thanos ruler and thanos querier.
+    queryConfigSecret: {
+      apiVersion: 'v1',
+      kind: 'Secret',
+      metadata: {
+        name: 'thanos-ruler-query-config',
+        namespace: cfg.namespace,
+        labels: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:210]: replace by commonLabels

File: jsonnet/thanos-ruler.libsonnet [hunk 11]
```diff
@@ -1,457 +1,458 @@
...
-    kind: 'ServiceAccount',
-    metadata: {
-      name: 'thanos-ruler',
-      namespace: cfg.namespace,
-      annotations: {
-        'serviceaccounts.openshift.io/oauth-redirectreference.thanos-ruler': '{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"thanos-ruler"}}',
+    serviceAccount: {
+      apiVersion: 'v1',
+      kind: 'ServiceAccount',
+      metadata: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:234]: add commonLabels

File: jsonnet/thanos-ruler.libsonnet [hunk 12]
```diff
@@ -1,457 +1,458 @@
...
-      namespace: cfg.namespace,
-      annotations: {
-        'service.beta.openshift.io/serving-cert-secret-name': 'thanos-ruler-tls',
+    service+: {
+      metadata+: {
+        name: 'thanos-ruler',
+        annotations: {
+          'service.beta.openshift.io/serving-cert-secret-name': 'thanos-ruler-tls',
+        },
+        labels: cfg.labels,
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:249]: we should use commonLabels here instead

File: jsonnet/thanos-ruler.libsonnet [hunk 13]
```diff
@@ -1,457 +1,458 @@
...
       },
     },
-    spec: {
-      selector: {
-        matchLabels: cfg.labels,
+
+    serviceMonitor+: {
+      metadata+: {
+        name: 'thanos-ruler',
+        labels: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:262]: use commonLabels

File: jsonnet/thanos-ruler.libsonnet [hunk 14]
```diff
@@ -1,457 +1,458 @@
...
-          port: 'web',
-          interval: '30s',
-          scheme: 'https',
-          tlsConfig: {
-            caFile: '/etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt',
-            serverName: 'server-name-replaced-at-runtime',
-          },
-          bearerTokenFile: '/var/run/secrets/kubernetes.io/serviceaccount/token',
+      spec+: {
+        selector+: {
```
Comment (dgrisonnet) [inline jsonnet/thanos-ruler.libsonnet:267]: use cfg.podLabelSelector

**Review by dgrisonnet — COMMENTED** [review]
I made a couple of comments regarding what we could leverage from the upstream library. Also, we should remove these custom labels https://github.com/openshift/cluster-monitoring-operator/blob/master/jsonnet/main.jsonnet#L281-L287 in favor of the upstream ones https://github.com/thanos-io/kube-thanos/blob/main/jsonnet/kube-thanos/kube-thanos-rule.libsonnet#L29-L40

File: jsonnet/thanos-ruler.libsonnet [hunk 15]
```diff
@@ -1,457 +1,458 @@
...
+          resources: ['securitycontextconstraints'],
+          verbs: ['use'],
+        },
+      ],
+    },
+
+    clusterRoleBinding: {
+      apiVersion: 'rbac.authorization.k8s.io/v1',
+      kind: 'ClusterRoleBinding',
+      metadata: {
```
[the remark this answers is not in the document]
Comment (raptorsun) [reply jsonnet/thanos-ruler.libsonnet:90]: does clusterRole need commonLablels,too ?

File: jsonnet/thanos-ruler.libsonnet (hunk 3 above)
Comment (raptorsun) [reply jsonnet/thanos-ruler.libsonnet]: clusterIP inherited from upstream, the 2 elements "sessionAffinity" and "type" are left as they are.

**Comment by dgrisonnet** [comment]
@paulfantom can you please have a look at this PR? I'd like to have a second pair of :eyes: on the inclusion of upstream thanos-ruler configuration.

**Comment by dgrisonnet** [comment]
The `ci/prow/e2e-agnostic-operator` failures are legit.

/hold

**Comment by raptorsun** [comment]
the label 'config.openshift.io/inject-trusted-cabundle': 'true' is brought back to trustedCaBundle for thanos-ruler.

**Comment by simonpasquier** [comment]
@raptorsun please hold on this PR. Hard affinity doesn't play well with persistent storage and it may break upgrades from 4.7 to 4.8 [1].

[1] https://bugzilla.redhat.com/show_bug.cgi?id=1967614

**Comment by raptorsun** [comment]
Pull request is closed for now, until we resolve bug [1967614](https://bugzilla.redhat.com/show_bug.cgi?id=1967614).

**Pull request closed without merging**